FLT/Mathlib/NumberTheory/Padics/PadicIntegers.lean

```diff
@@ -51,23 +52,124 @@ lemma isOpenEmbedding_coe : IsOpenEmbedding ((↑) : ℤ_[p] → ℚ_[p]) := by
 -- Yaël: Do we really want this as a coercion?
 noncomputable instance : Coe ℤ_[p]ˣ ℚ_[p]ˣ where coe := Units.map Coe.ringHom.toMonoidHom
 
-/-- For a `ℤ_[p]`-submodule `s` of `ℚ_[p]`, `x • s` has index `‖x‖⁻¹` in `s`.
+/-- `x • S` has index `‖x‖⁻¹` in `S`, where `S` is the copy of `ℤ_[p]` inside `ℚ_[p]` --/
+
+lemma smul_submodule_one_relindex (hx: x ≠ 0):
+    (x • (1 : Submodule ℤ_[p] ℚ_[p]).toAddSubgroup).relindex (1 : Submodule ℤ_[p] ℚ_[p]).toAddSubgroup = ‖x.val‖₊⁻¹ := by
+
+  -- We first prove a closely related statement using subgroups of `ℤ_[p]`. We then use the coercion from
+  -- ℤ_[p] to ℚ_[p] to obtain the desired result for the groups considered
+  -- as subgroups of `ℚ_[p]`.
+
+  let H := (x • (1 : Submodule ℤ_[p] ℤ_[p])).toAddSubgroup
+
+  have H_relindex_Z : (H.index) = ‖(x : ℚ_[p])‖₊⁻¹ := by
+    have x_factor := PadicInt.unitCoeff_spec hx
+    norm_cast at x_factor
+    unfold H
+    nth_rw 1 [x_factor]
+    conv =>
+      pattern _ • _
+      equals (Ideal.span {(p: ℤ_[p]) ^ (x.valuation)}) =>
+        ext y
+        refine ⟨?_, ?_⟩
+        . intro hy
+          rw [← Submodule.singleton_set_smul, Submodule.mem_singleton_set_smul] at hy
+          obtain ⟨m, ⟨_, hm⟩⟩ := hy
+          rw [Ideal.mem_span_singleton']
+          simp only [Nat.cast_pow, smul_eq_mul] at hm
+          rw [mul_comm, ← mul_assoc] at hm
+          exact ⟨(m * ↑(PadicInt.unitCoeff hx)), hm.symm⟩
+
+        . intro hy
+          simp
+          rw [← Submodule.singleton_set_smul]
+          rw [Submodule.mem_singleton_set_smul]
+          simp only [Submodule.mem_top, smul_eq_mul, true_and]
+          simp_rw [mul_assoc, mul_comm, ← mul_assoc, mul_comm]
+
+          rw [Ideal.mem_span_singleton'] at hy
+          obtain ⟨b, hb⟩ := hy
+          use (b * ↑(PadicInt.unitCoeff hx)⁻¹)
+          simp [hb.symm]
+
+    have quotient_equiv_zmod := RingHom.quotientKerEquivOfSurjective (f := PadicInt.toZModPow (x.valuation) (p := p)) (R := ℤ_[p]) (ZMod.ringHom_surjective (PadicInt.toZModPow x.valuation))
+    have card_eq_zmod := Nat.card_congr quotient_equiv_zmod.toEquiv
+
+    have quotient_equiv_quotient := Ideal.quotEquivOfEq (PadicInt.ker_toZModPow (x.valuation) (p := p))
+    have card_eq_quotient := Nat.card_congr quotient_equiv_quotient.toEquiv
+    rw [card_eq_quotient] at card_eq_zmod
+    have quotients_defeq: (ℤ_[p] ⧸ (Submodule.toAddSubgroup (Ideal.span {(p : ℤ_[p]) ^ x.valuation}))) = (ℤ_[p] ⧸ Ideal.span {(p: ℤ_[p]) ^ x.valuation}) := by
+      rfl
+
+    dsimp [AddSubgroup.index]
+    rw [quotients_defeq, card_eq_zmod]
+    dsimp [nnnorm]
+    rw [Subtype.ext_iff_val]
+    simp [PadicInt.norm_eq_zpow_neg_valuation hx]
```

Comment:
This is basically proving that the index of the ideal generated by p^n in Z_p is p^n. This should be a standalone lemma, not something proved in the middle of one gigantic theorem.